salt/metalk8s/orchestrate/upgrade/init.sls

```diff
@@ -92,6 +92,86 @@ Deploy node {{ node }}:
 
 {%- endfor %}
 
+{#- In MetalK8s-2.4.2 we added selector in nginx ingress controller DaemonSet
+    and nginx ingress default backend Deployment but `selector` are immutable
+    field so, in this case, we cannot replace the object we need to first
+    remove the current one and then deploy the desired one. #}
+{#- For upgrade we should use the current version to check whether or not we
+    need to remove the nginx ingress objects
+    `if current_version < 2.4.2`
+    but in orchestrate we are not able to known the current version so add an
+    hack checking if the current object exist and has a component selector, as
+    by default K8S add selector:
+    ```
+    matchLabels:
+      app: nginx-ingress
+      component: controller
+      release: nginx-ingress
+    ```
+    when our current selector is
+    ```
+    matchLabels:
+      app: nginx-ingress
+      release: nginx-ingress
+    ``` #}
+{%- set nginx_ingress_ds = salt.metalk8s_kubernetes.get_object(
+        kind='DaemonSet',
+        apiVersion='extensions/v1beta1',
+        name='nginx-ingress-controller',
+        namespace='metalk8s-ingress',
+        kubeconfig=kubeconfig,
+        context=context
+    ) %}
+{%- set nginx_ingress_deploy = salt.metalk8s_kubernetes.get_object(
+        kind='Deployment',
+        apiVersion='extensions/v1beta1',
+        name='nginx-ingress-default-backend',
+        namespace='metalk8s-ingress',
+        kubeconfig=kubeconfig,
+        context=context
+    ) %}
+{%- set current_selector = {
```

Comment:
`desired_selector` would be more explicit IMHO, since the "current" one sounds like the one used in the _DaemonSet_ and _Deployment_ objects you just retrieved.